### PR TITLE
Change `FromWire` to accept an `Arena` argument

### PR DESCRIPTION
**Pull request by mcy**
This change implements the design described in #34.

Note that this change adds a single block of `unsafe` in `mem.rs`. I'm not sure how I feel about this, and I kind of want to have an explicit policy for accepting new `unsafe`.

### CODE REVIEW

**Comment by mcy** [comment]
@osenft I'm not sure whether you want to review this change, but it's definitely very relevant to you, so seeing how it changes the way things are done might be a good idea.

**Comment by mcy** [comment]
@jrvanwhy would you mind reviewing `mem.rs` for correctness? (I can't seem to add you as a reviewer...)

**Comment by jrvanwhy** [comment]
> @jrvanwhy would you mind reviewing `mem.rs` for correctness? (I can't seem to add you as a reviewer...)

LGTM!

File: src/mem.rs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        }
+
+        self.cursor.set(proposed_cursor);
+        // At this point, it is not possible for following calls to access
+        // the range buf_ptr[cursor..proposed_cursor]. Moreover, the invariants
+        // of BumpArena ensure that buf_ptr + cursor must land inside (or one
+        // past) the allocation. Hence, the pointer arithmetic that follows is
+        // safe, as is the cast to isize.
+        let slice = unsafe {
+            // SAFE: buf_len >= cursor, and buf_len <= isize::MAX.
```
Comment (jon-flatley) [inline src/mem.rs:123]: nit: Should be `buf_len < isize::MAX`

**Pull request merged**